ui/src/main/java/org/pentaho/di/ui/core/widget/TableView.java

```diff
@@ -2283,32 +2300,65 @@ private void editCombo( TableItem row, int rownr, int colnr ) {
       }
     }
 
-    combo = new CCombo( table, colinfo.isReadOnly() ? SWT.READ_ONLY : SWT.NONE );
-    props.setLook( combo, Props.WIDGET_STYLE_TABLE );
-    combo.addTraverseListener( lsTraverse );
-    combo.addModifyListener( lsModCombo );
-    combo.addFocusListener( lsFocusCombo );
-
     String[] opt = getComboValues( row, colinfo );
-
     if ( colinfo.getComboValuesSelectionListener() != null ) {
       opt = colinfo.getComboValuesSelectionListener().getComboValues( row, rownr, colnr );
     }
-    combo.setItems( opt );
-    combo.setVisibleItemCount( opt.length );
-    combo.setText( row.getText( colnr ) );
-    if ( lsMod != null ) {
-      combo.addModifyListener( lsMod );
+
+    final boolean useVariables = colinfo.isUsingVariables();
+    if ( useVariables ) {
+      GetCaretPositionInterface getCaretPositionInterface = new GetCaretPositionInterface() {
+        @Override
+        public int getCaretPosition() {
+          return 0;
+        }
+      };
+
+      // The text widget will be disposed when we get here
+      // So we need to write to the table row
+      //
+      InsertTextInterface insertTextInterface = new InsertTextInterface() {
+        @Override
+        public void insertText( String string, int position ) {
+          StringBuilder buffer = new StringBuilder( table.getItem( rownr ).getText( colnr ) );
+          buffer.insert( position, string );
+          table.getItem( rownr ).setText( colnr, buffer.toString() );
+          int newPosition = position + string.length();
+          edit( rownr, colnr );
+        }
+      };
+
+      combo = new ComboVar( variables, table, SWT.SINGLE | SWT.LEFT | SWT.BORDER, getCaretPositionInterface, insertTextInterface );
+      ( (ComboVar) combo ).addModifyListener( lsModCombo );
```

Comment:
Can we avoid the repeated casting by creating a one local variable with the correct type?